components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -253,52 +253,46 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         store_ctx: &mut StoreContext<EK, ER, T>,
         req: RaftCmdRequest,
     ) {
-        match self.validate_commit_merge(&req) {
-            Some(true) if self.is_leader() => {
-                let (ch, _) = CmdResChannel::pair();
-                self.on_admin_command(store_ctx, req, ch);
-            }
-            Some(false) => {
-                let commit_merge = req.get_admin_request().get_commit_merge();
-                let source_id = commit_merge.get_source_state().get_region().get_id();
-                let _ = store_ctx.router.force_send(
-                    source_id,
-                    PeerMsg::RejectCommitMerge {
-                        index: commit_of_merge(commit_merge),
-                    },
-                );
-            }
-            _ => (),
-        }
-    }
-
-    fn validate_commit_merge(&self, req: &RaftCmdRequest) -> Option<bool> {
         let expected_epoch = req.get_header().get_region_epoch();
         let merge = req.get_admin_request().get_commit_merge();
         assert!(merge.has_source_state() && merge.get_source_state().has_merge_state());
         let source_region = merge.get_source_state().get_region();
         let region = self.region();
-        if self
+        if let Some(r) = self
             .storage()
             .region_state()
             .get_merged_records()
             .iter()
-            .any(|p| p.get_source_region_id() == source_region.get_id())
+            .find(|p| p.get_source_region_id() == source_region.get_id())
```

Comment:
We need to ensure the state won't be rollback by restart before acking.

Reply:
What do you mean? If it "restart before acking", source peer will not rollback, it will ask again and be acknowledged again.

Reply:
Supposing target region commit merge at index i. If target peer is started and decide to catch up logs before index i, then it's possible to execute commit merge again. If source peer is acked before restarted, then it may destroy itself already, hence target peer will either be blocked or panic.

Reply:
I still don't understand. Isn't this what we wanted? that source peer destroys itself even when target peer hasn't persisted its state. After restart target doesn't need to wait for source peer.

Reply:
> Supposing target region commit merge at index i. If target peer is started and decide to catch up logs before index i, then it's possible to execute commit merge again. If source peer is acked before restarted, then it may destroy itself already, hence target peer will either be blocked or panic.

@BusyJay 
Could you elaborate a bit?
So should the catch up logs happen before the source peer send AskCommitMerge request to its local target peer? Or do you mean today the catch up logs does not flush and thus after restart it needs to catch up again?

Reply:
Discussed offline.